Kaarina Mattika, CS87A, Section 1758
https://github.com/kmattika/cs82a-portfolio

In [1]:
import pandas as pd
import seaborn as sns

In [2]:
df = pd.read_csv(r'C:\Users\kamat\OneDrive\Desktop\CS82A\bridges.csv')

In [3]:
display(df.head(10))
display(df.shape)
display(df.info())
display(df.describe())

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
0,E1,M,3.0,1818,HIGHWAY,NaN,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
1,E2,A,25.0,1819,HIGHWAY,1037.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
2,E3,A,39.0,1829,AQUEDUCT,NaN,1.0,N,THROUGH,WOOD,NaN,S,WOOD
3,E5,A,29.0,1837,HIGHWAY,1000.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
4,E6,M,23.0,1838,HIGHWAY,NaN,2.0,N,THROUGH,WOOD,NaN,S,WOOD
5,E7,A,27.0,1840,HIGHWAY,990.0,2.0,N,THROUGH,WOOD,MEDIUM,S,WOOD
6,E8,A,28.0,1844,AQUEDUCT,1000.0,1.0,N,THROUGH,IRON,SHORT,S,SUSPEN
7,E9,M,3.0,1846,HIGHWAY,1500.0,2.0,N,THROUGH,IRON,SHORT,S,SUSPEN
8,E10,A,39.0,1848,AQUEDUCT,NaN,1.0,N,DECK,WOOD,NaN,S,WOOD
9,E11,A,29.0,1851,HIGHWAY,1000.0,2.0,N,THROUGH,WOOD,MEDIUM,S,WOOD


(108, 13)

<class 'pandas.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Id        108 non-null    str    
 1   river     108 non-null    str    
 2   location  107 non-null    float64
 3   erected   108 non-null    int64  
 4   purpose   108 non-null    str    
 5   length    81 non-null     float64
 6   lanes     92 non-null     float64
 7   clear-g   106 non-null    str    
 8   t-or-d    102 non-null    str    
 9   material  106 non-null    str    
 10  span      92 non-null     str    
 11  rel-l     103 non-null    str    
 12  type      106 non-null    str    
dtypes: float64(3), int64(1), str(9)
memory usage: 14.7 KB


None

,location,erected,length,lanes
count,107.000000,108.000000,81.000000,92.000000
mean,25.978505,1905.314815,1567.469136,2.630435
std,13.665871,37.173664,747.491523,1.164657
min,1.000000,1818.000000,804.000000,1.000000
25%,15.500000,1884.000000,1000.000000,2.000000
50%,27.000000,1903.000000,1300.000000,2.000000
75%,37.500000,1928.000000,2000.000000,4.000000
max,52.000000,1986.000000,4558.000000,6.000000


From the early snapshot provided by 'head', 'shape', 'info', and 'describe', I can draw some early hypothesies from these numbers. Using the most general statistical analysis, any value beyond three deviations from the mean has a likelihood of 0.1% chance of occuring - thus, any bridge length outside the bounds of -675.005 and 3809.943 is statistically improbable. As it is impossible to have a bridge with a negative length, that leaves us with only the positive third deviation value of 3809.943, of which there is at least one value in this dataset (that we are presently aware of) that is outside these bounds.

In [4]:
df[df['length'] >= 3809.944]

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
31,E34,O,41.0,1888,RR,4558.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
44,E46,A,37.0,1897,RR,4000.0,2.0,G,DECK,STEEL,LONG,F,SIMPLE-T


From our query, we have two entries with values in 'length' above 3809.943. These entries have relatively similar values in 'erected', and exact value matches in 'purpose', 'lanes', 'clear-g', 'material', 'span', 'rel-l', and 'type', with their differences coming in 'river', and t-or-d'. My plan to investigate these values is to compare each of these entries with entries that have similar periods in 'erected', exact matches in 'purpose', 'lanes', and 'material'. I do not find the categories of 'river', 'location', 'lanes', 'clear-g', 'span', 'rel-l', or 'type' to be germane to the ground truth behind the data. I will use one half standard deviations as the basis for what constitutes similar data.

In [5]:
df[(df['erected'] > 1869) & (df['erected'] < 1907) & (df['purpose'] == 'RR') & (df['lanes'] == 2.0) & (df['material'] == 'STEEL') & (df['span'] == 'LONG')]

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
31,E34,O,41.0,1888,RR,4558.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
44,E46,A,37.0,1897,RR,4000.0,2.0,G,DECK,STEEL,LONG,F,SIMPLE-T
49,E94,M,13.0,1901,RR,NaN,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T


For the first entry, 'E34', we can see that there are two similar matches: 'E46' and 'E94'. E46 is our other outlier, so we can discard that in our current analysis. Our second match, 'E94', does not have true data for it's own length.

In [6]:
df[(df['erected'] > 1878) & (df['erected'] < 1916) & (df['purpose'] == 'RR') & (df['lanes'] == 2.0) & (df['material'] == 'STEEL') & (df['span'] == 'LONG')]

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
31,E34,O,41.0,1888,RR,4558.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
44,E46,A,37.0,1897,RR,4000.0,2.0,G,DECK,STEEL,LONG,F,SIMPLE-T
49,E94,M,13.0,1901,RR,NaN,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
61,E57,O,49.0,1910,RR,1620.0,2.0,G,THROUGH,STEEL,LONG,F,CANTILEV
65,E61,O,41.0,1915,RR,2822.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T


Our second entry yields a few more matches. We can discard 'E34' and 'E94' for the reasons stated previously. That leaves us with two plausible alternatives, 'E57' and 'E61'. E57 was built later than E46, and has a significantly shorter length of 1620.0, 2380 units fewer. It also has a different 'type', noted as a cantilever type bridge. E61 was also built later than E46, and is also much shorter with a length of 2822, 1178 units fewer. E61 matches E46 in all other aspects, except for values in 'river' and 't-or-d'. The most critical values here are of length, and both bridges of similar construction to our outliers E34 and E46 have statistically significant differences in length.

In [7]:
df[df['length'] > 2314.96]

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
31,E34,O,41.0,1888,RR,4558.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
41,E42,M,9.0,1895,HIGHWAY,2367.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
44,E46,A,37.0,1897,RR,4000.0,2.0,G,DECK,STEEL,LONG,F,SIMPLE-T
52,E87,A,35.0,1903,RR,3000.0,2.0,G,THROUGH,STEEL,MEDIUM,S,SIMPLE-T
65,E61,O,41.0,1915,RR,2822.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
71,E66,A,32.0,1924,HIGHWAY,2365.0,4.0,G,THROUGH,STEEL,MEDIUM,S,ARCH
76,E72,M,5.0,1927,HIGHWAY,2663.0,4.0,N,DECK,STEEL,MEDIUM,S-F,CANTILEV
78,E75,A,30.0,1928,HIGHWAY,2678.0,4.0,G,DECK,STEEL,MEDIUM,F,ARCH
95,E81,M,14.0,1951,HIGHWAY,2423.0,4.0,G,DECK,STEEL,LONG,F,CONT-T
104,E91,O,44.0,1975,HIGHWAY,3756.0,6.0,G,THROUGH,STEEL,LONG,F,ARCH


We can use our heuristic skills here and examine the list of bridges with lengths of one standard deviation above the mean. Our outlier values, E34 and E46, only have a single other bridge of remotely similar length present. E91, which was constructed in 1975, with a length of 3756 units, consisting of six lanes, and uses an arch-type of construction, cannot plausibly support the presumption that the lengths of E34 and E46 are true. The differences in construction material and technique span nearly a century of time, and given that the positive third standard deviation of the mean has a value of 3809.943, E91 itself is just barely statistically plausible. Thus, I would remove rows 31 and 44 (corresponding to E34 and E46 respectively).

Data cleaning notes, in order:

Removed rows 31, 44, as we cannot trust that data.
Fill in null values found in 'length' with the column's median value.

In [8]:
df = df.drop([31, 44])
fill_value = df['length'].median()
df['length'] = df['length'].fillna(fill_value)

In [9]:
display(df.head(10))
display(df.shape)
display(df.info())
display(df.describe())

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
0,E1,M,3.0,1818,HIGHWAY,1245.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
1,E2,A,25.0,1819,HIGHWAY,1037.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
2,E3,A,39.0,1829,AQUEDUCT,1245.0,1.0,N,THROUGH,WOOD,NaN,S,WOOD
3,E5,A,29.0,1837,HIGHWAY,1000.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
4,E6,M,23.0,1838,HIGHWAY,1245.0,2.0,N,THROUGH,WOOD,NaN,S,WOOD
5,E7,A,27.0,1840,HIGHWAY,990.0,2.0,N,THROUGH,WOOD,MEDIUM,S,WOOD
6,E8,A,28.0,1844,AQUEDUCT,1000.0,1.0,N,THROUGH,IRON,SHORT,S,SUSPEN
7,E9,M,3.0,1846,HIGHWAY,1500.0,2.0,N,THROUGH,IRON,SHORT,S,SUSPEN
8,E10,A,39.0,1848,AQUEDUCT,1245.0,1.0,N,DECK,WOOD,NaN,S,WOOD
9,E11,A,29.0,1851,HIGHWAY,1000.0,2.0,N,THROUGH,WOOD,MEDIUM,S,WOOD


(106, 13)

<class 'pandas.DataFrame'>
Index: 106 entries, 0 to 107
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Id        106 non-null    str    
 1   river     106 non-null    str    
 2   location  105 non-null    float64
 3   erected   106 non-null    int64  
 4   purpose   106 non-null    str    
 5   length    106 non-null    float64
 6   lanes     90 non-null     float64
 7   clear-g   104 non-null    str    
 8   t-or-d    100 non-null    str    
 9   material  104 non-null    str    
 10  span      90 non-null     str    
 11  rel-l     101 non-null    str    
 12  type      104 non-null    str    
dtypes: float64(3), int64(1), str(9)
memory usage: 15.2 KB


None

,location,erected,length,lanes
count,105.000000,106.000000,106.000000,90.000000
mean,25.730476,1905.556604,1434.169811,2.644444
std,13.672883,37.478395,541.278495,1.173788
min,1.000000,1818.000000,804.000000,1.000000
25%,15.000000,1884.000000,1065.000000,2.000000
50%,27.000000,1903.500000,1245.000000,2.000000
75%,37.000000,1928.000000,1644.000000,4.000000
max,52.000000,1986.000000,3756.000000,6.000000
